freeglut/freeglut/include/GL/freeglut_ext.h

```diff
@@ -157,6 +157,7 @@ FGAPI void    FGAPIENTRY glutSetMenuFont( int menuID, void* font );
 /*
  * Window-specific callback functions, see fg_callbacks.c
  */
+FGAPI void    FGAPIENTRY glutKeyboardExtFunc( void (* callback)( int, int, int ) );
```

Comment:
Should we have a `glutKeyboardUpExtFunc`? For completeness?

Reply:
A character may be the result of several keypresses, such as the compose key or extended latin "dead key" input, and especially so when typing for example chinese text using an IME. As such, there is no corresponding 'up' event for a character event coming from the system. In windows the characters come from a WM_CHAR event, and on X11 the XFilterEvents function removes events eaten by the IME.

Reply:
Ok. Good to know.

Reply:
Which is why it's confusing to call this glutKeyboardExtFunc. A better name might be glutCharFunc similar to the glfw notation.

Reply:
If we're suggesting different names, why not be obvious: `glutKeyboardUnicodeFunc`?

Reply:
I picked the name because it was similar to the original behavior of glut where glutKeyboardFunc was used to detect characters, and glutSpecialFunc was used to detect presses of non-character keys. This was before glutKeyboardUpFunc was added to support games wanting to know key press events, and conflated the event of a keyboard key going down and up and the event of a character being generated.

As such, it may be better to call the new function glutCharacterFunc or something similar, to make clear that distinction... unfortunately glutKeyboardFunc applies the keyboard mapping and sends events on a key down event, with some special hacks to do the same mapping to KeyboardUp events. Backwards compatibility prevents us from changing the behavior to what I would have done instead: glutRawKeyDownFunc, glutRawKeyUpFunc, which take scan codes and make glutKeyboardFunc take a 32-bit unicode character rather than a 7-bit ascii character.

Reply:
Then perhaps glutUnicodeFunc?

Reply:
I agree, Char or Character in the name might be lead someone to think that it's another interface for the same thing. I had the same confusion while reading the glfw docs. I still don't like glutUnicodeFunc for some reason, but I'll aggree it's the best candidate so far. I'll post again if I can thing something better.

Reply:
Either glutUnicodeFunc or glutUnicodeCharFunc would work best for me i think, don't see other options. And this is higher level than the keyboard in our current scheme of things